mono/metadata/marshal.c

```diff
@@ -11180,7 +11181,8 @@ ves_icall_System_Runtime_InteropServices_Marshal_SizeOf (MonoReflectionTypeHandl
 		return 0;
 	}
 
-	return mono_class_native_size (klass, NULL);
+	result = mono_marshal_type_size (type, NULL, &align, FALSE, klass->unicode);
+	return (guint32)result;
```

Comment:
Why is this needed ?

Reply:
The native size is 0 for an empty struct and for a struct containing empty structs the size will also be transitively wrong

Reply:
Shouldn't the native size be 1 as well ?

Reply:
It's not, it's 0 everywhere in the runtime. It's 1 on .NET framework.

Reply:
The behavior is visible, and compatibility with mono and .net is impossible? Or behavior is only visible when marshaling? Persistant data compat?

Reply:
Do we have a test covering this behavior? If so I would be ok with merging it.

Reply:
The test I added covers this (the test won't pass without it) but I don't know how good our coverage is otherwise.

Reply:
I mean mono_class_native_size () should return the same value as mono_marshal_type_size (), mono_marshal_type_size () calls mono_class_native_size () for structs:

	case MONO_NATIVE_STRUCT: 
		klass = mono_class_from_mono_type (type);
		if (klass == mono_defaults.object_class &&
			(mspec && mspec->native == MONO_NATIVE_STRUCT)) {
		*align = 16;
		return 16;
		}
		return mono_class_native_size (klass, align);

Reply:
But `mono_marshal_type_size` has been updated at https://github.com/mono/mono/pull/7142/files#diff-559b8405966c917ff45da9a2a9bf0296R11942 to guarantee a size of at least 1.

Reply:
Wouldn't it be better to do this in mono_class_native_size () ?

Reply:
I don't know, maybe? It seems like `mono_class_native_size` is used in many more places than `mono_marshal_type_size` so I don't know what's the impact of such a change. How about we merge here and figure out what else we can transition from `mono_class_native_size` to `mono_marshal_type_size` in a follow-up PR, and possibly merge both.

Reply:
My original attempted fix was to make the native size of empty structs 1 byte, as in .net framework. It broke a bunch of things, in part because it seems like we have multiple copies of size/offset calculation logic in different parts of mono and I failed to find and fix all of them.